File: Exercise 1/MergeSort/MergeSort.py
from time import *

# Read Input
with open('input.txt', 'r') as file:
    lines = file.readlines()
    arrs = []
    for i in range(0, len(lines), 2):
        n = int(lines[i].strip())
        arrs.append(list(map(int, lines[i + 1].strip().split())))

# Merge Sort
def MergePart(arr : list, low : int, mid : int, high : int):
    i, j = low, mid + 1

    temp = []
    while i <= mid and j <= high:
        if arr[i] < arr[j]:
            nextVal = arr[i]
            i += 1
        else:
            nextVal = arr[j]
            j += 1

        temp.append(nextVal)

    while i <= mid:
        temp.append(arr[i])
        i += 1
    while j <= high:
        temp.append(arr[j])
        j += 1

    for i in range(len(temp)):
        arr[low + i] = temp[i]

def MergeSort(arr : list, low : int, high : int):
    if low >= high:
        return

    mid = (low + high) // 2

    MergeSort(arr, low, mid)
    MergeSort(arr, mid + 1, high)

    MergePart(arr, low, mid, high)


# Sort and Check time
for arr in arrs:
    start_time = time()
    MergeSort(arr, 0, len(arr) - 1)
    finish_time = time()

    print(f"Array (Sorted in {finish_time - start_time:.6f} seconds)")

File: Sorting Report/MergeSort/MergeSort.cpp
#include <iostream>
#include <vector>
#include <fstream>
#include <sstream>
#include <chrono>

using namespace std;
using namespace chrono;

void MergePart(vector<int>& arr, int low, int mid, int high) {
    int i = low, j = mid + 1;
    vector<int> temp;

    while (i <= mid && j <= high) {
        if (arr[i] < arr[j])
            temp.push_back(arr[i++]);
        else
            temp.push_back(arr[j++]);
    }

    while (i <= mid)
        temp.push_back(arr[i++]);
    while (j <= high)
        temp.push_back(arr[j++]);

    for (int k = 0; k < temp.size(); k++)
        arr[low + k] = temp[k];
}

void MergeSort(vector<int>& arr, int low, int high) {
    if (low >= high) return;

    int mid = (low + high) / 2;

    MergeSort(arr, low, mid);
    MergeSort(arr, mid + 1, high);
    MergePart(arr, low, mid, high);
}

int main() {
    ifstream inputFile("input.txt");
    vector<vector<int>> arrs;
    string line;

    // Read Input
    while (getline(inputFile, line)) {
        int n = stoi(line), num;
        vector<int> arr;

        if (getline(inputFile, line)) {
            stringstream ss(line);
            while (ss >> num)
                arr.push_back(num);
        }

        arrs.push_back(arr);
    }

    inputFile.close();

    // Sort and Check Time
    for (int i = 0; i < arrs.size(); i++) {
        auto start = high_resolution_clock::now();
        MergeSort(arrs[i], 0, arrs[i].size() - 1);
        auto end = high_resolution_clock::now();

        duration<double> exec_time = end - start;

        cout << "Array " << i + 1 << " (Sorted in " << exec_time.count() << " seconds)" << endl;
    }

    return 0;
}
